### DISCUSSION TITLE
Challenges with combining domain and non-domain states in xstate

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2405] [General]
Greetings everyone.
Fantastic library and community.
Thank you very much indeed.

I am struggling for a week now trying to get my head around some interesting challenges that actually are not specific to xstate but are rather general state management problems. Nevertheless, they take specific forms and shapes when are attempted to be solved using xstate.

As we all know, applications are comprised of different kind of states:
- Domain state which is the core of the application and is usually persisted server-side and cached client-side. It is made of all the business entities and their graph-like relationships.
- Non-domain state which is everything else including state of UI components, UI router location, I/O such as networking, etc. It is usually transient at client-side, but not always where in this case would be more appropriately considered as domain state.

Given this context in mind, the UI is expected to render domain and non-domain states accordingly. For this, I am trying to leverage xstate to manage both end-to-end, while keeping my UI components free of stateful logic. I am actually passing one xstate service for each stateful UI component, then I bind the component with the service state/context and send UI events back to the service. This will allow me to implement and test all application state logic at the service layer, decoupled from any UI implementation at the presentation layer. The service APIs become the interface contracts between the two layers. So far so good.

Now in order to achieve these design goals, I need a mechanism to combine both domain and non-domain states within xstate services. Let's look into the nature of each with respect to xstate:
- **Domain state:**
  1. It is received from the server-side using invoked/spawned services;
  2. It takes the shape of a hierarchy where each node is a specific business entity usually with a unique ID;
  3. It mostly represents the extended state (i.e. context) of a service's state machine;
  4. It may contains some flag/status fields that can be represented by the service's finite states.
- **Non-domain state:**
  1. It is generated at the client-side;
  2. It is usually either standalone state (i.e. singletons such as networking/API clients) or augmented state (i.e. transients such as UI element selection/position) to business entities of domain state;
  3. It mostly represents the finite state of a service's state machine;
  4. It may contains some data fields that can be represented by the service's extended state (i.e. context).

After this long background, the specific challenges that I am facing are actually specific to combining domain state with _augmented_ non-domain state in xstate services:

Normally, augmented non-domain state can be extracted in generic and reusable xstate services. Now, the question is how can I reuse (i.e. combine) these services within the domain state xstate services? For this we have two approaches:
  1. **Merging the definitions and options of the non-domain state services into the domain state services.** The only problem here is how can we then resolve naming conflicts between the merged context fields, states, events, actions, guards, etc. without sorting to prefixes that jeopardize the readability of the code? Examples of such long prefixes are SOME_BUSINESS_ENTITY_EVENT_X and SELECTION_EVENT_Y. There is actually another problem with this approach as it cannot solve the same issue with third-party services that do not usually follow our naming convention. Also, xstate currently do not support the use of JS symbols to uniquely identity these concepts. Moreover, I do not know of an easy and DRY way to allow for strongly typed renaming of such concepts at merging time.
  2. **Composing the actual implementations of the non-domain state services with the domain state services.** This could actually solve the problem, and the produced deep nesting can be mitigated by flatting the spawned actor services inside the consumer UI components using the `useActor` hook. However, this also introduces new questions:
      - When the domain state has no finite states, do we need an xstate service just to hold the domain state and the references to the non-domain state spawned actor services (i.e. the augumented finite state) where they will become the children of a top-level service? 
      - Should we actually use xstate services when the domain state has some fields that can be represented using finite states? Or should we rely on xstate guards to control the state transitions with good old equality comparisons against such fields?

Previous to xstate, I used to have the domain and non-domain states separated into two different stores: the caching layer (e.g. using URQL caching) and the presentation layer (e.g. using Vue reactive composition API). Now with xstate, I am hoping to centralize all state needs into one service layer independent form other implementation details.

Looking forward to hear from everybody about their approaches and experiences on how to go about solving these challenges.

**Comment by ChrisShank** [comment]
Curious, have you found the distinction between domain and non-domain state useful?

**Reply by geekox86** [reply]
Hi Chris, thank you for passing by.

It is only useful in defining the challenge and scope of the problem, as I noticed that it only manifests with domain state that needs to be combined with non-domain state.

Also, previous to xstate, separating the two kinds in different stores was my typical solution.

Though, now depending on the suggested solution with xstate, this distinction may not be necessary?

Do you manage both kinds in xstate?
What is your general approach especially with xstate service reuse?

**Reply by ChrisShank** [reply]
I will answer, might be a long one 🙃

**Comment by a-y-u-s-h** [comment]
I was just discussing something like this with a friend but I am having hard time understanding the challenges that you've mentioned without an example state machine (can you provide any so that I can think over it with you? Thanks). But, because we both are trying to solve the same problem in our own ways, I will describe my approach below (work in progress, everything remaining is a challenge that I would like to add to this thread) -

If I wanted to create a general Saas, I wouldn't start right away by creating the business logic state charts in a state node. I would instead break it into features and then label the business logic node as layer 2 solution which contains these layer 1 charts. Let's say I want to have a state chart of a whatsapp clone - I wouldn't name the states or events or actors based on business, but rather what that feature actually is - it's chat networking. I would first keep a general `application` state machine and then abstract out the actors later. Like you, I would keep the non-domain information separate (which I can arrange in several ways, but I mainly have 4 parallel states `theme`, `layout`, `content` and `document`). `document` state has all DOM events attached, and every event looks sort of like this: 

```yaml
on:
  mouse.press:
    - cond: target.container
      actions: document.container.mouse.press
    - cond: target.registration.form.submit.button
      actions: document.registration.form.submit.mouse.press
```

Don't mind the dotted naming of actions, I have gotten used to it. Also love them because they're super readable. Anyway, I usually make hierarchy from the top. I have some top level actors like `user`, `database`, `interface`, ...and rest I'm figuring out still, could be related to payment methods, etc. 

![screenshot_20210715_025238](https://user-images.githubusercontent.com/19586719/125695285-4515e650-9286-430d-9048-7a262e43ccef.png)

Next, I keep all user related information within user. In case I am building some application like a Zoom clone where the server can get multiple users interacting with each other, I would extract users out as actors but usually while development I don't. In a general Saas app, they have their `authentication`, `authorization` all inside them. 

![screenshot_20210715_024353](https://user-images.githubusercontent.com/19586719/125696044-fca349fd-8e0b-468f-9097-f12278089f2e.png)

And depending on what kind of authorization level they have, they affect the permissions on database layer as well though I know this should just be handled on backend. Just playing around on frontend as well, because I may want to show different interfaces altogether to different kinds of users (administrators, developers, individuals, teams, etc.).

![screenshot_20210715_025251](https://user-images.githubusercontent.com/19586719/125696224-bd7364ed-eca7-4564-ba9c-f6efbb86445a.png)



Have to admit that making clean interface state machines is indeed very difficult and I have gone through several approaches (I have described one [here](https://github.com/statelyai/xstate/discussions/1633#discussioncomment-129375) as well), but I am going to experiment soon by simply making an event bus and following some convention for events related to interface. I've taken a week off from work to work on my state modelling fu, make these general app features so that I don't have to make these ever again. If something miraculous happens and I figure it all out, I will release it as a video course or something.
